packages/caver-wallet/src/keyring/keyring.js

```diff
@@ -0,0 +1,845 @@
+     * @return {Keyring}
+     */
+    static recover(message, signature, preFixed = false) {
+        if (_.isObject(message)) {
+            return this.recover(message.messageHash, message.signature, true)
+        }
+
+        if (!preFixed) {
+            message = utils.hashMessage(message)
+        }
+
+        return AccountLib.recover(message, AccountLib.encodeSignature(signature)).toLowerCase()
+    }
+
+    /**
+     * creates a keyring.
+     * @param {string} address - The address of keyring.
+     * @param {string|Array.<string>|Array.<Array<string>>|PrivateKey|Array.<PrivateKey>|Array.<Array<PrivateKey>>} key - The key(s) to use in keyring.
+     */
+    constructor(address, key) {
+        this.address = address
+        this.key = key
+    }
+
+    /**
+     * @type {string}
+     */
+    get address() {
+        return this._address
+    }
+
+    set address(addressInput) {
+        if (!utils.isAddress(addressInput)) throw new Error(`Invalid address : ${addressInput}`)
+
+        this._address = utils.addHexPrefix(addressInput).toLowerCase()
+    }
+
+    /**
+     * @type {Array.<PrivateKey>}
+     */
+    get roleTransactionKey() {
+        return this.getKeyByRole(KEY_ROLE.ROLE_TRANSACTION_KEY)
+    }
+
+    /**
+     * @type {Array.<PrivateKey>}
+     */
+    get roleAccountUpdateKey() {
+        return this.getKeyByRole(KEY_ROLE.ROLE_ACCOUNT_UPDATE_KEY)
+    }
+
+    /**
+     * @type {Array.<PrivateKey>}
+     */
+    get roleFeePayerKey() {
+        return this.getKeyByRole(KEY_ROLE.ROLE_FEE_PAYER_KEY)
+    }
+
+    /**
+     * @type {Array.<Array.<PrivateKey>>}
+     */
+    get key() {
+        return this._key
+    }
+
+    set key(keyInput) {
+        this._key = formattingForKeyInKeyring(keyInput)
+    }
+
+    /**
+     * returns public key strings in format of role-based.
+     *
+     * @return {Array.<Array<string>>}
+     */
+    getPublicKey() {
+        const publicKeys = generateKeysFormat()
+        for (let i = 0; i < KEY_ROLE.ROLE_LAST; i++) {
+            for (const k of this._key[i]) {
+                publicKeys[i].push(k.getPublicKey())
+            }
+        }
+        return publicKeys
+    }
+
+    /**
+     * returns a copied keyring instance
+     *
+     * @return {Keyring}
+     */
+    copy() {
+        return new Keyring(this._address, this._key)
+    }
+
+    /**
+     * signs with transactionHash with key and returns signature.
+     *
+     * @param {string} transactionHash The hash of transaction.
+     * @param {string|number} chainId The chainId or the network.
+     * @param {number} role A number indicating the role of the key. You can use `caver.wallet.keyring.role`.
+     * @param {number} [index] The index of the key to be used. (default: 0)
+     * @return {Array<string>}
+     */
+    signWithKey(transactionHash, chainId, role, index = 0) {
+        if (!utils.isTxHashStrict(transactionHash)) throw new Error(`Invalid transaction hash: ${transactionHash}`)
```

Comment:
Hmm... The function name is wierd. At the first look, I think isTxHashStrict() determines whether the tx hash can be found in the network. Is there any difference of hash and txhash for this function?

Reply:
'isTxHash' and 'isTxHashStrict' only check the length of the hash.
Therefore, it does not actually check if there are any transactions on the network.
Since it is already published, it seems difficult to rename it.
What about creating another function with 'isValidHash' or something else and using it internally?

Reply:
@jimni1222 good. If it is already published, please remain as it is. As you suggested, please make another function for this. Also please add a comment that the function will be deprecated so that users do not use it in the future anymore.

Reply:
I added 'isValidHash' and 'isValidHashStrict' functions!